src/network/wait-online/wait-online.c

```diff
@@ -81,18 +87,66 @@ static int parse_argv(int argc, char *argv[]) {
                 case ARG_VERSION:
                         return version();
 
-                case 'i':
-                        if (strv_extend(&arg_interfaces, optarg) < 0)
+                case 'i': {
+                        _cleanup_free_ char *ifname = NULL;
+                        LinkOperationalState s;
+                        const char *p;
+
+                        p = strchr(optarg, ':');
+                        if (p) {
+
+                                if (isempty(p + 1))
+                                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                               "Operational state is empty.");
+
+                                s = link_operstate_from_string(p + 1);
+                                if (s < 0)
+                                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                               "Invalid operational state '%s'", p + 1);
+
+                                ifname = strndup(optarg, p - optarg);
+                        } else {
+                                s = _LINK_OPERSTATE_INVALID;
+                                ifname = strdup(optarg);
+                        }
+                        if (!ifname)
                                 return log_oom();
 
-                        break;
+                        if (!ifname_valid(ifname))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "Invalid interface name '%s'", ifname);
+
+                        r = hashmap_ensure_allocated(&arg_interfaces, &string_hash_ops);
+                        if (r < 0)
+                                return log_oom();
+
+                        r = hashmap_put(arg_interfaces, ifname, INT_TO_PTR(s));
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to store interface name: %m");
+                        if (r == 0)
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "Interface name %s is already specified", ifname);
```

Comment:
I think it'd make sense to split the parsing into it's own function. It's a bit too much to put in a case stmt.